### ISSUE TITLE
Use a singleton pattern for included common/split chunks and modules

### ISSUE THREAD

**Issue by karolyi** [webpack/webpack#14113]
## Feature request

**What is the expected behavior?**

Generated chunks should use a singleton pattern so as to not execute on each subsequent inclusion.

**What is motivation or use case for adding/changing the behavior?**

Coming from https://github.com/django-webpack/django-webpack-loader/issues/289, the generated modules tree by webpack-bundle-tracker for JS modules in django templates will include the split chunks for each and every module that uses them. Upon using (rendering) more than one module on a page, the rendered common/split chunks will be executed many times, which is a bad thing: they might have initialization code that can cause trouble, as it does in my case. They should only execute once. For details, see the linked issue.

**How should this be implemented in your opinion?**

Split/common modules (if not all) should use a singleton pattern to not be executed again upon multiple inclusion.

**Are you willing to work on this yourself?**

yes, but I'm not a JS/webpack expert, hence I'm not sure as to where to make the changes. if you can point it out exactly, I might be able to implement a workaround.

**Comment by alexander-akait** [comment]
Sorry, it is expected and js always run again when you override something

**Comment by alexander-akait** [comment]
Only for ECMA modules it is valid

**Comment by karolyi** [comment]
Not sure I understood your answer, much less the reason why you closed this so abruptly.

Would you please elaborate.

**Comment by alexander-akait** [comment]
What in my answer needs to be clarified? if you add:
```
<script type="text/javascript" src="http://localhost:8001/one.js"></script>
<script type="text/javascript" src="http://localhost:8001/two.js"></script>
```

**one.js**

```js
var foo = 1;
```

**two.js**

```js
var foo = 2;
```

`window.foo` === `2`

**Comment by karolyi** [comment]
The problem is, I'm talking about split modules that are split by webpack itself. Those modules should not execute on each inclusion, in my opinion. Those should execute once and only once on each inclusion, or not be included for every module that uses them, since django will include them for each rendered module on the same page.

**Comment by alexander-akait** [comment]
> The problem is, I'm talking about split modules that are split by webpack itself. Those modules should not execute on each inclusion, in my opinion.

Sorry not, it is violate spec, if you need to do it, you can do it on own application side or create custom plugin for this, anyway it is wrong usage.

**Comment by karolyi** [comment]
What spec does it violate, can you cite it to me please?

All I need is, more control over the split modules. Interestingly, this wasn't an issue with CommonChunksPlugin or whatever it was called in webpack V3. It only became an issue in V4 and upwards.

**Comment by alexander-akait** [comment]
> What spec does it violate, can you cite it to me please?

Browser execute code twice when you use `text/javascript`.

If you need another logic you can create custom webpack plugin for this. Also singleton code increase size of runtime and most of developers don't need this (and again it is wrong), also this is big breaking change and we don't want to do it in core.

**Comment by karolyi** [comment]
The logic is broken too when one uses webpack-bundle-tracker and expects that when using more than one entry points in a page, a split module used by both would run only once.

Question is, why is not wanting to add a small addition to webpack (having control over split modules to _optionally_ execute only once) override another breakage, that is the aforementioned common/split modules executing as many times as included because of webpack-bundle-tracker including them for each module.

I hope you can see my issue here.

**Comment by alexander-akait** [comment]
Because it is out of scope default behavior, we have plugin api for exotic usaage

**Comment by karolyi** [comment]
So let me get this straight:

webpack, a module bundler framework defines it to be 'out of scope' behavior to provide a functionality that is logical for any other language: included modules should only run on first inclusion, and use their previously initialized reference afterwards.

I guess I give up, then.

**Comment by alexander-akait** [comment]
This is how `type="text/javascript"` works in browser, migrate on ecma modules and use `type="module"` and you will get necessary behavior

**Comment by karolyi** [comment]
I haven't used `type="module"` so far but will check it out, thanks. Browser support for `type="module"` is also a question here.

Problem is, I don't have control over the output of `django-webpack-loader` and the json output of `webpack-bundle-tracker`. Even if I start adding this functionality to them, it will most probably end up with an option added back to webpack, to be able to make a distinction between split modules (`type="module"`) and entry points (`type="text/javascript"`).

Just my hunch so far, this theory still needs testing.

**Comment by karolyi** [comment]
Linking for reference: https://github.com/django-webpack/webpack-bundle-tracker/issues/31

**Comment by karolyi** [comment]
For anyone interested, we decided to go with the optional skipping of multiple outputs of shared chunks in `django-webpack-loader`:

https://github.com/django-webpack/django-webpack-loader/pull/297
